src/Lean/Meta/Tactic/Rewrites.lean

```diff
@@ -0,0 +1,378 @@
+    let (_, _, type) ← forallMetaTelescopeReducing (← inferType h)
+    let type ← whnfR type
+    match type.getAppFnArgs with
+    | (``Eq, #[_, lhs, rhs])
+    | (``Iff, #[lhs, rhs]) => do
+      let lhsKey : Array DiscrTree.Key ← DiscrTree.mkPath lhs discrTreeConfig
+      let rhsKey : Array DiscrTree.Key ← DiscrTree.mkPath rhs discrTreeConfig
+      result := result.push (h, false, forwardWeight * lhsKey.size)
+        |>.push (h, true, backwardWeight * rhsKey.size)
+    | _ => pure ()
+  return result
+
+/--
+We drop `.star` and `Eq * * *` from the discriminator trees because
+they match too much.
+-/
+def droppedKeys : List (List LazyDiscrTree.Key) := [[.star], [.const `Eq 3, .star, .star, .star]]
+
+def createModuleTreeRef : MetaM (LazyDiscrTree.ModuleDiscrTreeRef (Name × Bool × Nat)) :=
+  LazyDiscrTree.createModuleTreeRef addImport droppedKeys
+
+private def ExtState := IO.Ref (Option (LazyDiscrTree (Name × Bool × Nat)))
+
+private builtin_initialize ExtState.default : IO.Ref (Option (LazyDiscrTree (Name × Bool × Nat))) ← do
+  IO.mkRef .none
+
+private instance : Inhabited ExtState where
+  default := ExtState.default
+
+private builtin_initialize ext : EnvExtension ExtState ←
+  registerEnvExtension (IO.mkRef .none)
+
+/--
+The maximum number of constants an individual task may perform.
+
+The value was picked because it roughly correponded to 50ms of work on the
+machine this was developed on.  Smaller numbers did not seem to improve
+performance when importing Std and larger numbers (<10k) seemed to degrade
+initialization performance.
+-/
+private def constantsPerImportTask : Nat := 6500
+
+def incPrio : Nat → Name × Bool × Nat → Name × Bool × Nat
+| p, (nm, d, prio) => (nm, d, prio * 100 + p)
+
+/-- Create function for finding relevant declarations. -/
+def rwFindDecls (moduleRef : LazyDiscrTree.ModuleDiscrTreeRef (Name × Bool × Nat)) : Expr → MetaM (Array (Name × Bool × Nat)) :=
+  LazyDiscrTree.findMatchesExt moduleRef ext addImport
+      (droppedKeys := droppedKeys)
+      (constantsPerTask := constantsPerImportTask)
+      (adjustResult := incPrio)
+
+/-- Data structure recording a potential rewrite to report from the `rw?` tactic. -/
+structure RewriteResult where
+  /-- The lemma we rewrote by.
+  This is `Expr`, not just a `Name`, as it may be a local hypothesis. -/
+  expr : Expr
+  /-- `True` if we rewrote backwards (i.e. with `rw [← h]`). -/
+  symm : Bool
+  /-- The "weight" of the rewrite. This is calculated based on how specific the rewrite rule was. -/
+  weight : Nat
+  /-- The result from the `rw` tactic. -/
+  result : Meta.RewriteResult
+  /-- Pretty-printed result. -/
+  -- This is an `Option` so that it can be computed lazily.
+  ppResult? : Option String
+  /-- The metavariable context after the rewrite.
+  This needs to be stored as part of the result so we can backtrack the state. -/
+  mctx : MetavarContext
+
+/-- Update a `RewriteResult` by filling in the `rfl?` field if it is currently `none`,
```

Comment:
I see you've removed the cache for the `computeRfl` result in `RewriteResult`. I presume that's reasonable, but could you explain? This doc-string will need updating.

Reply:
As far as I can tell, both the `rfl?` nor `ppResult?` fields cached values that were only read once, as such it didn't seem worth caching them.

Are there users of this datatype downstream where caching would be useful?  I can re-add if needed.